### PR TITLE
[@types/ember] add ember__template

### PR DESCRIPTION
**Pull request by scottkidder**
Moved htmlSafe and isHTMLSafe into @ember/template where the official docs say they are.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Review by mike-north — APPROVED** [review]
LGTM

File: types/ember__template/index.d.ts
Comment (jamescdavis) [inline types/ember__template/index.d.ts]: @scottkidder, wanna put your info here?

File: types/ember__template/index.d.ts
Comment (jamescdavis) [inline types/ember__template/index.d.ts]: IMHO, we should move `@ember/string/-private/handlebars`  to *this* package (`@ember/template/-private/handlebars`).

**Review by jamescdavis — COMMENTED** [review]
Looks pretty good, but I think we should move the `SafeString` type def here.

**Comment by jamescdavis** [comment]
Also, see: https://github.com/typed-ember/ember-cli-typescript/issues/784#issuecomment-534840744

File: types/ember__template/index.d.ts
Comment (scottkidder) [reply types/ember__template/index.d.ts]: Ok, I'll go ahead and do that :+1:

**Comment by scottkidder** [comment]
@typescript-bot Please don't merge this yet

**Comment by scottkidder** [comment]
@jamescdavis I think I've properly moved `-template/handlebars.d-ts` into `@ember/template`. I'm not sure if the `@deprecated` comment on the exports in string are actually working.

File: types/ember__string/index.d.ts
Comment (jamescdavis) [inline types/ember__string/index.d.ts]: Instead of redefining these here, IMHO it's better to re-export them from `@ember/template`:
```ts
export { htmlSafe, isHTMLSafe } from '@ember/template';
```
That way we keep a single definition for these functions and you don't have to import `SafeString` from `@ember/template/-private` here. I'm not sure if the deprecated annotation will be picked up by tslint doing it this way, but I'd still add it for documentation purposes.

Comment (scottkidder) [reply types/ember__string/index.d.ts]: Yeah I've done that now. I agree, from what I could find I don't think `@deprecated` will picked up this way.

**Review by jamescdavis — APPROVED** [review]
Looks great!

File: types/ember/index.d.ts
Comment (mike-north) [inline types/ember/index.d.ts]:
```suggestion
import * as EmberTemplateHandlebarsNs from '@ember/template/-private/handlebars';
```

File: types/ember__string/index.d.ts
Comment (mike-north) [inline types/ember__string/index.d.ts]: I'm unclear here -- is this

(a) a bug fix in our types, where `import { htmlSafe } from '@ember/template';` no longer works entirely in Ember 3.x

or 

(b) A deprecation in Ember 3.x that currently works, but logs a deprecation warning?

**Review by mike-north — CHANGES_REQUESTED** [review]

File: types/ember__template/index.d.ts
Comment (mike-north) [inline types/ember__template/index.d.ts]: While anyone can add themselves to these headers in DT, you should be aware of what this means
- your own code changes will be merged in automatically _without a review_, and _in 24 hours, instead of the usual 5-7 days_
- your approvals to incoming PRs will cause them to automatically be merged, and then new releases of these packages will be immediately cut and released.

If things go wrong as a result of your code changes and/or approvals, are you willing to dig in on short notice and untangle things so we can unblock the community?

Comment (scottkidder) [reply types/ember__template/index.d.ts]: Hi Mike, given that I'm still novice with TypeScript I think it's better that I remove myself from the headers. Thanks for the information!

File: types/ember__string/index.d.ts
Comment (scottkidder) [reply types/ember__string/index.d.ts]: @mike-north as far as I can tell `import { htmlSafe } from '@ember/string';` still works and is not deprecated. However official [docs](http://api.emberjs.com/ember/3.13/functions/@ember%2Ftemplate/htmlSafe) say to import from `@ember/template`.

I imagine that the import from `@ember/string` should/will be deprecated at some point.

When I used `@deprecated` here I was thinking that I was deprecating the dts import itself. I'm happy to remove `@deprecated` to revisit at a later date since there is no underlying deprecation to match.

Comment (jamescdavis) [reply types/ember__string/index.d.ts]: Those were my findings as well. I guess remove `@deprecated`. Here's where they changed the docs: https://github.com/emberjs/ember.js/pull/16887 (not much info there)

File: types/ember__template/index.d.ts
Comment (jamescdavis) [reply types/ember__template/index.d.ts]: @mike-north should we add the Ember TS team so you're not solely responsible?

**Review by jamescdavis — CHANGES_REQUESTED** [review]
In`types/ember/index.d.ts`under `namespace Handlebars` need to:
```diff
- class SafeString extends EmberStringHandlebarsNs.SafeString {}
+ class SafeString extends EmberTemplateHandlebarsNs.SafeString {}
```

File: types/ember__string/index.d.ts
Comment (jamescdavis) [reply types/ember__string/index.d.ts]: Here we go: https://github.com/emberjs/rfcs/blob/master/text/0236-deprecation-ember-string.md

Comment (mike-north) [reply types/ember__string/index.d.ts]: We need to align with Ember's public API. If it's not deprecated there, we should not indicate anything is deprecated here.

File: types/ember__template/index.d.ts
Comment (mike-north) [reply types/ember__template/index.d.ts]: @jamescdavis that sounds good to me. Mind opening a separate PR to do this?

**Comment by scottkidder** [comment]
Thanks @jamescdavis! I apologize for the delay here..

File: types/ember__string/index.d.ts
Comment (scottkidder) [reply types/ember__string/index.d.ts]: Got it, makes sense. Thoughts on if it deserves another RFC to deprecate the exports from `@ember/string`?

Comment (jamescdavis) [reply types/ember__string/index.d.ts]: It's unclear whether
> 1. Move` htmlSafe` and `isHTMLSafe` to `@ember/template`

(from [the above RFC](https://github.com/emberjs/rfcs/blob/master/text/0236-deprecation-ember-string.md))
would include deprecating the exports from `@ember/string`. Regardless, the current state appears to be that importing from `@ember/string` not yet officially deprecated, just no longer being taught in the guides. Therefore, I agree with @mike-north that we shouldn't indicate deprecation here at this time.

File: types/ember__template/index.d.ts
Comment (jamescdavis) [reply types/ember__template/index.d.ts]: Happy to once these get merged.

**Review by jamescdavis — APPROVED** [review]
Looks great! Thanks @scottkidder!

**Review by mike-north — APPROVED** [review]
Looks great, thanks!

**Comment by RyanCavanaugh** [comment]
Good to merge then?

**Comment by mike-north** [comment]
> Good to merge then?

Yes

**Pull request merged**
